File: _includes/code/howto/configure-rq/update-collections-rq.py
# Code examples for updating collections with RQ-8 compression

# ==============================
# =====  CONNECT =====
# ==============================

# START ConnectCode
import os, weaviate

# Best practice: store your credentials in environment variables
weaviate_url = os.environ["WEAVIATE_URL"]
weaviate_api_key = os.environ["WEAVIATE_API_KEY"]

client = weaviate.connect_to_weaviate_cloud(
    cluster_url=weaviate_url, auth_credentials=weaviate_api_key
)
# END ConnectCode

from weaviate.classes.config import Configure

client.collections.delete("MyUncompressedCollection")
client.collections.create(
    "MyUncompressedCollection",
    vector_config=Configure.Vectors.text2vec_openai(
        quantizer=Configure.VectorIndex.Quantizer.none()
    ),
)

client.collections.delete("MyLegacyCollection")
client.collections.create(
    "MyLegacyCollection",
    vectorizer_config=Configure.Vectorizer.text2vec_weaviate(),
)

# ==============================
# =====  UPDATE SINGLE COLLECTION (HNSW) =====
# ==============================

# START UpdateSingleCollectionHNSW
from weaviate.classes.config import Reconfigure

collection = client.collections.get("MyUncompressedCollection")
collection.config.update(
    vector_config=Reconfigure.Vectors.update(
        name="default",
        vector_index_config=Reconfigure.VectorIndex.hnsw(
            quantizer=Reconfigure.VectorIndex.Quantizer.rq(bits=8),
        ),
    )
)
# END UpdateSingleCollectionHNSW

"""TODO[g-despot] Can't test until cluster has async indexing.
client.collections.delete("MyUncompressedCollection")
client.collections.create(
    "MyUncompressedCollection",
    vector_config=Configure.Vectors.text2vec_openai(
        vector_index_config=Configure.VectorIndex.dynamic(),
        quantizer=Configure.VectorIndex.Quantizer.none(),
    ),
)
# ==============================
# =====  UPDATE SINGLE COLLECTION (DYNAMIC) =====
# ==============================

# START UpdateSingleCollectionDynamic
from weaviate.classes.config import Reconfigure

# For dynamic indexes, only the HNSW portion can be updated after creation
# The flat index compression settings are immutable
collection = client.collections.get("MyUncompressedCollection")
collection.config.update(
    vector_config=Reconfigure.Vectors.update(
        name="default",
        vector_index_config=Reconfigure.VectorIndex.dynamic(
            hnsw=Reconfigure.VectorIndex.hnsw(
                quantizer=Reconfigure.VectorIndex.Quantizer.rq(bits=8),
            ),
        ),
    )
)
# END UpdateSingleCollectionDynamic
"""
# ==============================
# =====  UPDATE LEGACY COLLECTION (pre-named vectors) =====
# ==============================

# START UpdateLegacyCollection
from weaviate.classes.config import Reconfigure

# For collections created before named vectors were introduced (pre-v1.24),
# use vector_index_config directly instead of vector_config
collection = client.collections.get("MyLegacyCollection")
collection.config.update(
    vector_index_config=Reconfigure.VectorIndex.hnsw(
        quantizer=Reconfigure.VectorIndex.Quantizer.rq(bits=8),
    )
)
# END UpdateLegacyCollection

# ==============================
# =====  LIST COLLECTIONS BY INDEX TYPE =====
# ==============================

# START ListCollectionsByIndexType
from weaviate.collections.classes.config import (
    _VectorIndexConfigHNSW,
    _VectorIndexConfigFlat,
    _VectorIndexConfigDynamic,
)

# Group collections by their vector index type
hnsw_collections = []
flat_collections = []
dynamic_collections = []
legacy_collections = []

collections = client.collections.list_all()

for collection_name in collections:
    collection = client.collections.get(collection_name)
    config = collection.config.get()

    # Check if this is a legacy collection (no named vectors)
    if not config.vector_config:
        # Legacy collection - check the top-level vector_index_config
        legacy_collections.append(collection_name)
        continue

    # For each named vector, determine its index type
    for vector_name, vector_config in config.vector_config.items():
        index_config = vector_config.vector_index_config
        entry = {"collection": collection_name, "vector": vector_name}

        if isinstance(index_config, _VectorIndexConfigHNSW):
            hnsw_collections.append(entry)
        elif isinstance(index_config, _VectorIndexConfigFlat):
            flat_collections.append(entry)
        elif isinstance(index_config, _VectorIndexConfigDynamic):
            dynamic_collections.append(entry)

print(f"HNSW collections: {len(hnsw_collections)}")
print(f"Flat collections: {len(flat_collections)}")
print(f"Dynamic collections: {len(dynamic_collections)}")
print(f"Legacy collections: {len(legacy_collections)}")
# END ListCollectionsByIndexType

# ==============================
# =====  UPDATE MULTIPLE COLLECTIONS =====
# ==============================

# START UpdateMultipleCollections
from weaviate.classes.config import Reconfigure

# Process collections in batches to avoid cluster instability
BATCH_SIZE = 100

# Only process the first batch (adjust slice for subsequent batches)
batch = hnsw_collections[:BATCH_SIZE]

for entry in batch:
    collection_name = entry["collection"]
    vector_name = entry["vector"]

    collection = client.collections.get(collection_name)
    print(f"Enabling RQ-8 compression for {collection_name} (vector: {vector_name})")
    # END UpdateMultipleCollections
    """
    # START UpdateMultipleCollections
    collection.config.update(
        vector_config=Reconfigure.Vectors.update(
            name=vector_name,
            vector_index_config=Reconfigure.VectorIndex.hnsw(
                quantizer=Reconfigure.VectorIndex.Quantizer.rq(bits=8),
            ),
        )
    )
    # END UpdateMultipleCollections
    """
    # START UpdateMultipleCollections

print(
    f"Processed {len(batch)} collections. Remaining: {len(hnsw_collections) - BATCH_SIZE}"
)
# END UpdateMultipleCollections

# ==============================
# =====  CHECK COMPRESSION STATUS =====
# ==============================

# START CheckCompressionStatus
collection = client.collections.get("MyUncompressedCollection")
config = collection.config.get()

# Check if this is a legacy collection (no named vectors)
if config.vector_config:
    # Named vectors - iterate through vector_config
    for vector_name, vector_config in config.vector_config.items():
        print(f"\nVector: {vector_name}")
        quantizer = vector_config.vector_index_config.quantizer
        if quantizer:
            print(f"  Quantizer type: {type(quantizer).__name__}")
            if hasattr(quantizer, "bits"):
                print(f"  Bits: {quantizer.bits}")
        else:
            print("  No compression enabled")
else:
    # Legacy collection - check vector_index_config directly
    print(f"\nLegacy collection (no named vectors)")
    quantizer = config.vector_index_config.quantizer
    if quantizer:
        print(f"  Quantizer type: {type(quantizer).__name__}")
        if hasattr(quantizer, "bits"):
            print(f"  Bits: {quantizer.bits}")
    else:
        print("  No compression enabled")
# END CheckCompressionStatus

client.close()

File: _includes/code/howto/configure-rq/update-collections-rq.ts
// Code examples for updating collections with RQ-8 compression

// ==============================
// =====  CONNECT =====
// ==============================

// START ConnectCode // START UpdateSingleCollectionHNSW // START UpdateSingleCollectionDynamic // START UpdateLegacyCollection // START ListCollectionsByIndexType // START UpdateMultipleCollections // START CheckCompressionStatus
import weaviate, { reconfigure } from 'weaviate-client';
// END ConnectCode // END UpdateSingleCollectionHNSW // END UpdateSingleCollectionDynamic // END UpdateLegacyCollection // END ListCollectionsByIndexType // END UpdateMultipleCollections // END CheckCompressionStatus

// START ConnectCode
// Best practice: store your credentials in environment variables
const weaviateUrl = process.env.WEAVIATE_URL!
const weaviateApiKey = process.env.WEAVIATE_API_KEY!

const client = await weaviate.connectToWeaviateCloud(weaviateUrl, {
    authCredentials: new weaviate.ApiKey(weaviateApiKey),
})
// END ConnectCode

import { configure } from 'weaviate-client';

await client.collections.delete("MyUncompressedCollection")
await client.collections.create({
    name: "MyUncompressedCollection",
    vectorizers: configure.vectors.text2VecOpenAI({
        quantizer: configure.vectorIndex.quantizer.none(),
    }),
})

/*TODO[g-despot] TypeScript client v3.x doesn't support creating true legacy collections (pre-v1.24 without named vectors).
The client internally uses named vectors for all collections. The UpdateLegacyCollection snippet below is still valid
for actual legacy collections in clusters, it just can't be tested in this file.
await client.collections.delete("MyLegacyCollection")
await client.collections.create({
    name: "MyLegacyCollection",
    vectorizerConfig: configure.vectorizer.text2VecWeaviate(),
})
*///

// ==============================
// =====  UPDATE SINGLE COLLECTION (HNSW) =====
// ==============================

// START UpdateSingleCollectionHNSW

const collection = client.collections.get("MyUncompressedCollection")
await collection.config.update({
    vectorizers: [ reconfigure.vectors.update({
        name: "default",
        vectorIndexConfig: reconfigure.vectorIndex.hnsw({
            quantizer: reconfigure.vectorIndex.quantizer.rq({ bits: 8 }),
        }),
    })]
})
// END UpdateSingleCollectionHNSW

/*TODO[g-despot] Can't test until cluster has async indexing.
await client.collections.delete("MyUncompressedCollection")
await client.collections.create({
    name: "MyUncompressedCollection",
    vectorizers: configure.vectors.text2VecOpenAI({
        vectorIndexConfig: configure.vectorIndex.dynamic(),
        quantizer: configure.vectorIndex.quantizer.none(),
    }),
})
// ==============================
// =====  UPDATE SINGLE COLLECTION (DYNAMIC) =====
// ==============================

// START UpdateSingleCollectionDynamic

// For dynamic indexes, only the HNSW portion can be updated after creation
// The flat index compression settings are immutable
const collectionDynamic = client.collections.get("MyUncompressedCollection")
await collectionDynamic.config.update({
    vectorizers: [ reconfigure.vectors.update({
        name: "default",
        vectorIndexConfig: reconfigure.vectorIndex.dynamic({
            hnsw: reconfigure.vectorIndex.hnsw({
                quantizer: reconfigure.vectorIndex.quantizer.rq({ bits: 8 }),
            }),
        }),
    })]
})
// END UpdateSingleCollectionDynamic
*/
// ==============================
// =====  UPDATE LEGACY COLLECTION (pre-named vectors) =====
// ==============================

// START UpdateLegacyCollection

// For collections created before named vectors were introduced (pre-v1.24),
// use vectorIndexConfig directly instead of vectorizers
const legacyCollection = client.collections.get("MyLegacyCollection")
await legacyCollection.config.update({
    vectorIndexConfig: reconfigure.vectorIndex.hnsw({
        quantizer: reconfigure.vectorIndex.quantizer.rq({ bits: 8 }),
    })
})
// END UpdateLegacyCollection

// ==============================
// =====  LIST COLLECTIONS BY INDEX TYPE =====
// ==============================

// START ListCollectionsByIndexType

// Group collections by their vector index type
const hnswCollections: Array<{collection: string, vector: string}> = []
const flatCollections: Array<{collection: string, vector: string}> = []
const dynamicCollections: Array<{collection: string, vector: string}> = []
const legacyCollections: string[] = []

const allCollections = await client.collections.listAll()

for (const collectionConfig of Object.values(allCollections)) {
    const collectionName = collectionConfig.name
    try {
        const coll = client.collections.get(collectionName)
        const config = await coll.config.get()

        // Check if this is a legacy collection (no named vectors)
        if (!config.vectorizers || Object.keys(config.vectorizers).length === 0) {
            legacyCollections.push(collectionName)
            continue
        }

        // For each named vector, determine its index type
        for (const [vectorName, vectorConfig] of Object.entries(config.vectorizers)) {
            const indexType = vectorConfig.indexType
            const entry = { collection: collectionName, vector: vectorName }

            if (indexType === 'hnsw') {
                hnswCollections.push(entry)
            } else if (indexType === 'flat') {
                flatCollections.push(entry)
            } else if (indexType === 'dynamic') {
                dynamicCollections.push(entry)
            }
        }
    } catch (error) {
        console.log(`Skipping collection ${collectionName}: ${error}`)
    }
}

console.log(`HNSW collections: ${hnswCollections.length}`)
console.log(`Flat collections: ${flatCollections.length}`)
console.log(`Dynamic collections: ${dynamicCollections.length}`)
console.log(`Legacy collections: ${legacyCollections.length}`)
// END ListCollectionsByIndexType

// ==============================
// =====  UPDATE MULTIPLE COLLECTIONS =====
// ==============================

// START UpdateMultipleCollections

// Process collections in batches to avoid cluster instability
const BATCH_SIZE = 100

// Only process the first batch (adjust slice for subsequent batches)
const batch = hnswCollections.slice(0, BATCH_SIZE)

for (const entry of batch) {
    const collectionName = entry.collection
    const vectorName = entry.vector

    const collection = client.collections.get(collectionName)
    console.log(`Enabling RQ-8 compression for ${collectionName} (vector: ${vectorName})`)
    // END UpdateMultipleCollections
    /*
    // START UpdateMultipleCollections
    await collection.config.update({
        vectorizers: [ reconfigure.vectors.update({
            name: vectorName,
            vectorIndexConfig: reconfigure.vectorIndex.hnsw({
                quantizer: reconfigure.vectorIndex.quantizer.rq({ bits: 8 }),
            }),
        })]
    })
    // END UpdateMultipleCollections
    */
    // START UpdateMultipleCollections
}

console.log(`Processed ${batch.length} collections. Remaining: ${hnswCollections.length - BATCH_SIZE}`)
// END UpdateMultipleCollections

// ==============================
// =====  CHECK COMPRESSION STATUS =====
// ==============================

// START CheckCompressionStatus

const collectionToCheck = client.collections.get("MyUncompressedCollection")
const configToCheck = await collectionToCheck.config.get()

// Check if this is a legacy collection (no named vectors)
if (configToCheck.vectorizers && Object.keys(configToCheck.vectorizers).length > 0) {
    // Named vectors - iterate through vectorizers
    for (const [vectorName, vectorConfig] of Object.entries(configToCheck.vectorizers)) {
        console.log(`\nVector: ${vectorName}`)
        const quantizer = vectorConfig.indexConfig?.quantizer
        if (quantizer) {
            console.log(`  Quantizer type: ${quantizer.constructor.name}`)
            if ('bits' in quantizer) {
                console.log(`  Bits: ${quantizer.bits}`)
            }
        } else {
            console.log("  No compression enabled")
        }
    }
} else {
    // Legacy collection - check vectorIndexConfig directly
    console.log(`\nLegacy collection (no named vectors)`)
    const quantizer = configToCheck.vectorIndexConfig?.quantizer
    if (quantizer) {
        console.log(`  Quantizer type: ${quantizer.constructor.name}`)
        if ('bits' in quantizer) {
            console.log(`  Bits: ${quantizer.bits}`)
        }
    } else {
        console.log("  No compression enabled")
    }
}
// END CheckCompressionStatus

client.close()
